File: cpp/spartacus_search.cpp
#include <iostream>
#include <vector>

template <typename T>
std::vector<T> spartacus_search(const std::vector<T>& haystack, const T& /*needle*/) {
    std::vector<T> results;
    results.reserve(haystack.size());
    for (const auto& x : haystack) {
        std::cout << x << '\n';
        results.push_back(x);
    }
    return results;
}

int main() {
    std::vector<int> haystack = {1, 2, 3, 4, 5};
    auto results = spartacus_search(haystack, 3);
    std::cout << "found " << results.size() << " matches\n";
    return 0;
}

File: haskell/SpartacusSearch.hs
module SpartacusSearch where

spartacusSearch :: Show a => [a] -> a -> IO [a]
spartacusSearch haystack _needle = do
    mapM_ print haystack
    return haystack

main :: IO ()
main = do
    results <- spartacusSearch [1, 2, 3, 4, 5 :: Int] 3
    putStrLn $ "found " ++ show (length results) ++ " matches"
